# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/interfaces/base/nsIServiceWorkerManager.idl
# Commit: fb720435215f
# Full Hash: fb720435215fa5be3b2bd2101f3a54382954b70c
# Author: Andrew Sutherland <asutherland@asutherland.org>
# Date: 2024-10-24 09:44:34
# Regressor Bug: 1544232
# File Overlap Count: 4
# Description:
#   Bug 1544232 - Limit lifetime extension of SWs by SWs to the sender's lifetime. r=edenchuang
#   
#   This patch introduces an explicit concept of lifetimes with mechanisms
#   in place so that actions taken by Clients (windows or non-ServiceWorker
#   orkers) will extend the lifetime of a ServiceWorker, but a ServiceWorker
# ==============================================================================

diff -r 25452f6623a0 -r fb720435215f dom/interfaces/base/nsIServiceWorkerManager.idl
--- a/dom/interfaces/base/nsIServiceWorkerManager.idl	Thu Oct 24 03:02:41 2024 +0000
+++ b/dom/interfaces/base/nsIServiceWorkerManager.idl	Thu Oct 24 03:02:42 2024 +0000
@@ -54,6 +54,12 @@
   readonly attribute AString scriptSpec;
   readonly attribute AString cacheName;
 
+  // How many times has this ServiceWorker been launched since the registration
+  // was loaded?  This value is not persistent and starts at 0 every time the
+  // browser restarts.  The value also inherently starts at 0 for a freshly
+  // installed or updated ServiceWorker.
+  readonly attribute unsigned long launchCount;
+
   readonly attribute unsigned short state;
 
   readonly attribute nsIWorkerDebugger debugger;
@@ -67,6 +73,15 @@
   readonly attribute PRTime activatedTime;
   readonly attribute PRTime redundantTime;
 
+  // Returns the lifetime deadline of the ServiceWorker as the number of
+  // milliseconds since the (parent) process was created or 0 if there is no
+  // current deadline.  This is primarily intended to allow tests to compare
+  // consistent values, not be useful in general.  But note that
+  // `Components.utils.now()` operates in the same units (milliseconds since
+  // current process startup), and so can be used to hackily translate to wall
+  // clock time.
+  readonly attribute double lifetimeDeadline;
+
   // Total number of navigation faults experienced by this ServiceWorker since
   // it was loaded from disk at startup or was installed.
   readonly attribute unsigned long navigationFaultCount;
@@ -80,6 +95,17 @@
   void attachDebugger();
 
   void detachDebugger();
+
+  // Forcibly terminate the ServiceWorker if it is running, returning a promise
+  // that will be resolved when the worker is fully terminated.  If the
+  // ServiceWorker was not running, the promise will be resolved immediately.
+  //
+  // For the purposes of the ServiceWorkerManager, it's as if the ServiceWorker
+  // is already dead when the call completes and returns the Promise.  Any new
+  // functional events dispatched at the ServiceWorker will result in a new
+  // instance/global being spun up.
+  [implicit_jscontext]
+  Promise terminateWorker();
 };
 
 [scriptable, uuid(87e63548-d440-4b8a-b158-65ad1de0211E)]